File: HW1/MST.py
class Graph:
    def __init__(self, vertices):
        '''
        Constructor to initialize the graph with the number of vertices
        :param vertices: number of vertices in the graph
        '''
        self.V = vertices # Number of vertices
        self.graph = [] # List to store the edges

    def add_edge(self, u, v, w):
        '''
        Function to add an edge to the graph
        :param u: one side of the edge
        :param v: other side of the edge
        :param w: weight of the edge(cost to traverse the edge
        :return: None
        '''
        self.graph.append([u, v, w]) # Add the edge to the graph

    def find(self, parent, i): # disjoint set find operation
        '''
        Function to find the set of an element i (using path compression technique)
        :param parent: parent array to keep track of the parent of the set
        :param i: element whose set is to be found
        :return: parent of the set to which the element belongs
        '''
        if parent[i] == i: # If the element is the parent of itself
            return i
        return self.find(parent, parent[i]) # Recursively find the parent of the element

    def union(self, parent, rank, x, y): # disjoint set union operation
        '''
        Function to perform union of two sets x and y (x and y are the sets to which the elements belong)
        :param parent: parent array to keep track of the parent of the set
        :param rank: rank array to keep track of the rank of the set
        :param x: one of the sets
        :param y: other set
        :return: None
        '''
        xroot = self.find(parent, x) # Find the set to which x belongs
        yroot = self.find(parent, y) # Find the set to which y belongs

        if rank[xroot] < rank[yroot]: # If the rank of x is less than the rank of y
            parent[xroot] = yroot # Make y the parent of x
        elif rank[xroot] > rank[yroot]: # If the rank of x is greater than the rank of y
            parent[yroot] = xroot # Make x the parent of y
        else: # If the rank of x is equal to the rank of y
            parent[yroot] = xroot  # Make x the parent of y
            rank[xroot] += 1 # Increment the rank of x

    def kruskal_mst(self):
        '''
        Function to find the minimum spanning tree of the graph using Kruskal's algorithm
        :return: weight of the minimum spanning tree
        '''
        result = [] # List to store the edges of the minimum spanning tree
        i, e = 0, 0 # i = index of the edge, e = number of edges in the minimum spanning tree
        self.graph = sorted(self.graph, key=lambda item: item[2]) # Sort the edges based on their weights in non-decreasing order
        parent = [] # List to store the parent of the set to which the element belongs
        rank = [] # List to store the rank of the set to which the element belongs

        for node in range(self.V):
            parent.append(node) # Initially, the parent of each element is the element itself
            rank.append(0) # Initially, the rank of each set is 0

        while e < self.V - 1: # While the number of edges in the minimum spanning tree is less than the number of vertices - 1
            u, v, w = self.graph[i] # Get the edge with the minimum weight
            i = i + 1 # Move to the next edge
            x = self.find(parent, u) # Find the set to which u belongs
            y = self.find(parent, v) # Find the set to which v belongs

            if x != y: # If u and v belong to different sets (i.e., adding the edge does not form a cycle)
                e = e + 1 # Increment the number of edges in the minimum spanning tree
                result.append([u, v, w]) # Add the edge to the minimum spanning tree
                self.union(parent, rank, x, y) # Perform union of the sets to which u and v belong

        return sum(item[2] for item in result) # Return the weight(cost) of the minimum spanning tree


def main():
    '''
    Main function to take input and call the kruskal_mst function
    :return: None
    '''
    n, m = map(int, input().split()) # n = number of vertices, m = number of edges
    g = Graph(n) # Create a graph with n vertices
    for _ in range(m): # Add m edges to the graph
        u, v, w = map(int, input().split()) # u, v = vertices, w = weight
        g.add_edge(u - 1, v - 1, w) # Add edge to the graph
    print(g.kruskal_mst()) # Print the weight of the minimum spanning tree


if __name__ == "__main__": # If the script is run directly
    main() # Call the main function

File: HW1/MST.c
#include <stdio.h> // printf, scanf
#include <stdlib.h> // malloc, free, qsort

struct Edge {
    /* u, v: vertices of the edge
     * w: weight of the edge
     */
    int u, v, w;
};

struct Graph {
    /* V: number of vertices
     * E: number of edges
     * edge: array of edges of the graph (pointer to the first element of the array)
     */
    int V, E;
    struct Edge* edge;
};

struct Graph* createGraph(int V, int E) {
    /* Allocates memory for a graph with V vertices and E edges
     * and returns a pointer to the graph
     */
    struct Graph* graph = (struct Graph*) malloc(sizeof(struct Graph)); // Allocates memory for the graph
    graph->V = V; // Sets the number of vertices
    graph->E = E; // Sets the number of edges
    graph->edge = (struct Edge*) malloc(graph->E * sizeof(struct Edge));
    return graph; // Returns a pointer to the graph
}

int find(int parent[], int i) {
    /* disjoint set find operation
     * Finds the set of the element i
     * and performs path compression
     * using disjoint set data structure
     */
    if (parent[i] == i) // If i is the root
        return i; // Returns i
    return find(parent, parent[i]); // Else returns the root of the set of i with recursive call
}

void Union(int parent[], int rank[], int x, int y) {
    /* disjoint set union operation
     * Unites the sets of x and y
     * using disjoint set data structure
     */
    int xroot = find(parent, x); // Finds the set of x
    int yroot = find(parent, y); // Finds the set of y

    if (rank[xroot] < rank[yroot]) // If the rank of xroot is less than the rank of yroot
        parent[xroot] = yroot; // Then yroot becomes the parent of xroot
    else if (rank[xroot] > rank[yroot]) // If the rank of xroot is greater than the rank of yroot
        parent[yroot] = xroot; // Then xroot becomes the parent of yroot
    else { // If the ranks are the same
        parent[yroot] = xroot; // Then yroot becomes the parent of xroot
        rank[xroot]++; // And the rank of xroot increases by 1
    }
}

int compare(const void* a, const void* b) {
    /* Comparison function for qsort
     * Compares the weights of two edges
     */
    struct Edge* a1 = (struct Edge*)a; // Casts a to a pointer to an Edge
    struct Edge* b1 = (struct Edge*)b; // Casts b to a pointer to an Edge
    return a1->w > b1->w; // Returns 1 if the weight of a is greater than the weight of b, 0 otherwise
}

int kruskalMST(struct Graph* graph) {
    /* Kruskal's algorithm for finding the minimum spanning tree of a graph
     * Returns the weight of the minimum spanning tree
     * using disjoint set data structure and qsort
     */
    int V = graph->V; // Number of vertices
    struct Edge result[V]; // Array of edges of the minimum spanning tree
    qsort(graph->edge, graph->E, sizeof(graph->edge[0]), compare); // Sorts the edges of the graph by weight using qsort(quicksort)

    int* parent = (int*) malloc(V * sizeof(int)); // Array of parents of the vertices
    int* rank = (int*) malloc(V * sizeof(int)); // Array of ranks of the vertices
    for (int v = 0; v < V; ++v) { // Initializes the arrays
        parent[v] = v; // The parent of each vertex is itself
        rank[v] = 0; // The rank of each vertex is 0
    }

    int e = 0; // Number of edges of the minimum spanning tree
    int i = 0; // Index of the edge being considered
    while (e < V - 1 && i < graph->E) { // While the minimum spanning tree has less than V - 1 edges and there are edges left
        struct Edge next_edge = graph->edge[i++]; // The next edge to be considered
        int x = find(parent, next_edge.u); // The set of the first vertex of the edge
        int y = find(parent, next_edge.v); // The set of the second vertex of the edge

        if (x != y) { // If the vertices are not in the same set
            result[e++] = next_edge; // The edge is added to the minimum spanning tree
            Union(parent, rank, x, y); // The sets of the vertices are united
        }
    }

    int minCost = 0; // Weight of the minimum spanning tree
    for (i = 0; i < e; ++i) // Calculates the weight of the minimum spanning tree
        minCost += result[i].w; // Adds the weight of the edge to the weight of the minimum spanning tree
    return minCost; // Returns the weight of the minimum spanning tree
}

int main() {
    /* Reads the input and calls the kruskalMST function
     * Prints the weight of the minimum spanning tree
     */
    int V, E; // Number of vertices and edges
    scanf("%d %d", &V, &E); // Reads the number of vertices and edges

    struct Graph* graph = createGraph(V, E); // Creates a graph with V vertices and E edges

    for (int i = 0; i < E; ++i) { // Reads the edges
        scanf("%d %d %d", &graph->edge[i].u, &graph->edge[i].v, &graph->edge[i].w); // Reads the vertices and weight of the edge
        graph->edge[i].u--; // Vertices are 0-indexed (0 <= u, v < V)
        graph->edge[i].v--; // Vertices are 0-indexed (the reason for the decrement)
    }

    printf("%d\n", kruskalMST(graph)); // Calls the kruskalMST function and prints the weight of the minimum spanning tree

    return 0; // Returns 0 to the operating system
}